File: shorpypaper.py
#!/usr/local/bin/python3
from pyquery import PyQuery as pq
import requests
import subprocess
import time

tmpShorpyPhoto = '/tmp/dailyshorpy{}.jpg'.format(int(time.time()))

APPLESCRIPT = """/usr/bin/osascript<<END
    tell application "System Events"
        set desktopCount to count of desktops
        repeat with desktopNumber from 1 to desktopCount
            tell desktop desktopNumber
                set picture to "{}"
            end tell
        end repeat
    end tell
END"""


def main():

    # Load main site.
    root = 'http://www.shorpy.com'
    r = requests.get(root)
    j = pq(r.content)

    # Load first photo.
    first_photo = root + j('div.node div.content a').eq(1).attr('href')
    r = requests.get(first_photo)
    j = pq(r.content)
    image = j('img').eq(0).attr('src')

    # To reset the cached dailyshorpy.jpg.
    subprocess.Popen(APPLESCRIPT.format(
        '/Library/Desktop Pictures/Solid Colors/Solid Gray Light.png'),
        shell=True)

    with open(tmpShorpyPhoto, 'wb') as handle:
        request = requests.get(image, stream=True)
        for block in request.iter_content(1024):
            if not block:
                break
            handle.write(block)
        subprocess.Popen(APPLESCRIPT.format(tmpShorpyPhoto),
                         shell=True)


if __name__ == '__main__':
    main()

File: shorpypaper.js
#!/opt/homebrew/bin/node

const fs = require('fs');
const { execSync } = require('child_process');

const tmpShorpyPhoto = `/tmp/dailyshorpy${Math.floor(Date.now() / 1000)}.jpg`;

const APPLESCRIPT = `/usr/bin/osascript<<END
    tell application "System Events"
        set desktopCount to count of desktops
        repeat with desktopNumber from 1 to desktopCount
            tell desktop desktopNumber
                set picture to "{}"
            end tell
        end repeat
    end tell
END`;

async function main() {
  try {
    const root = 'http://www.shorpy.com';
    const mainPage = await fetch(root).then(r => r.text());

    // Find photo page links (div.node div.content a hrefs)
    const photoLinks = [...mainPage.matchAll(/div class="node"[\s\S]*?div class="content"[\s\S]*?<a href="([^"]+)"/g)];
    if (!photoLinks.length) throw new Error('No photo links found');
    const href = photoLinks[0][1];
    const firstPhotoLink = href.startsWith('http') ? href : root + href;

    // Load photo page, find the main photo (in /files/images/)
    const photoPage = await fetch(firstPhotoLink).then(r => r.text());
    const imgMatch = photoPage.match(/src="([^"]+\/files\/images\/[^"]+)"/);
    if (!imgMatch) throw new Error('No image found');
    // Get full-size by removing .preview from filename
    const imageUrl = imgMatch[1].replace('.preview', '');

    // Reset desktop
    execSync(APPLESCRIPT.replace('{}', '/Library/Desktop Pictures/Solid Colors/Solid Gray Light.png'),
      { shell: true });

    // Download and set wallpaper
    const imageData = await fetch(imageUrl).then(r => r.arrayBuffer());
    fs.writeFileSync(tmpShorpyPhoto, Buffer.from(imageData));
    execSync(APPLESCRIPT.replace('{}', tmpShorpyPhoto), { shell: true });
  } catch (error) {
    console.error('Error:', error.message);
  }
}

main();
